src/Npgsql/Internal/PgWriter.cs

```diff
@@ -35,31 +35,35 @@ sealed class NpgsqlBufferWriter : IStreamingWriter<byte>
     public void Advance(int count)
     {
         if (_lastBufferSize < count || _buffer.WriteSpaceLeft < count)
-            throw new InvalidOperationException("Cannot advance past the end of the current buffer.");
+            ThrowHelper.ThrowInvalidOperationException("Cannot advance past the end of the current buffer.");
         _lastBufferSize = null;
         _buffer.WritePosition += count;
     }
 
     public Memory<byte> GetMemory(int sizeHint = 0)
     {
-        if (sizeHint > _buffer.WriteSpaceLeft)
-            throw new OutOfMemoryException("Not enough space left in buffer.");
+        var writePosition = _buffer.WritePosition;
+        if (sizeHint > writePosition)
+            ThrowOutOfMemoryException();
 
-        var bufferSize = _buffer.WriteSpaceLeft;
+        var bufferSize = _buffer.Size - writePosition;
         _lastBufferSize = bufferSize;
-        return _buffer.Buffer.AsMemory(_buffer.WritePosition, bufferSize);
+        return _buffer.Buffer.AsMemory(writePosition, bufferSize);
     }
 
     public Span<byte> GetSpan(int sizeHint = 0)
     {
-        if (sizeHint > _buffer.WriteSpaceLeft)
-            throw new OutOfMemoryException("Not enough space left in buffer.");
+        var writePosition = _buffer.WritePosition;
+        if (sizeHint > writePosition)
+            ThrowOutOfMemoryException();
 
-        var bufferSize = _buffer.WriteSpaceLeft;
+        var bufferSize = _buffer.Size - writePosition;
         _lastBufferSize = bufferSize;
-        return _buffer.Buffer.AsSpan(_buffer.WritePosition, bufferSize);
+        return _buffer.Buffer.AsSpan(writePosition, bufferSize);
     }
 
+    static void ThrowOutOfMemoryException() => throw new OutOfMemoryException("Not enough space left in buffer.");
```

Comment:
We probably shouldn't ever throw oom, it's going to confuse both users and us. InvalidOperationException I think is fine.

Reply:
OOM is the contract of IBufferWriter
<img width="396" alt="Screenshot 2023-11-02 at 22 02 50" src="https://github.com/npgsql/npgsql/assets/4218809/c52f1d5c-302a-4058-b34c-6d059528e97b">

Reply:
🐝